### DISCUSSION TITLE
PrismaClientRustPanicError: Following a POST request to our application's endpoint from different origin.

### DISCUSSION THREAD

**Discussion by maxjnq** [prisma/prisma#23439] [Q&A]
Hello,

We're developing a straightforward waitlist tool designed to enable our users to integrate a signup form directly onto their own websites. Users embed a simple HTML code snippet that users can copy from our application, located at https://waitless.cc. Here is the HTML code snippet from our embed form.

```
<div class=container-__ID__ style="display: flex">
	<form
		autocomplete="off"
		class="form-__ID__"
		style="display: flex"
		action="__DOMAIN__/action/create-sub/__LISTID__"
		method="POST"
	>
		<input
			type="text"
			style="display: inline-flex"
			class="input-__ID__"
			name="email"
			type="email"
			placeholder="__PLACEHOLDER__"
			required
		/>

		<button type="submit" class="button-__ID__">
			<div class="cta-__ID__" style="opacity: __OCTA__; display: block">
				__CTA__
			</div>
			<div class="error-__ID__" style="opacity: 0; display: none"></div>
			<div class="loading-__ID__" style="opacity: __OLOADING__">
				<div class="loading-bar-__ID__">
					<div class="loading-pill-__ID__"></div>
				</div>
			</div>
		</button>
	</form>
</div>
```

The issue occurs when a POST request is sent from another origin (from the embed form above) to our Remix route /action/create-sub/$listId below:

```
import { parseWithZod } from '@conform-to/zod'
import { invariantResponse } from '@epic-web/invariant'
import {
	json,
	type ActionFunctionArgs,
	type HeadersFunction,
} from '@remix-run/node'
import { nanoid } from 'nanoid'
import { z } from 'zod'
import { EmailSchema } from '#app/components/design/public-form'
import { prisma } from '#app/utils/db.server'
import { getEnv } from '#app/utils/env.server'
import { checkHoneypot } from '#app/utils/honeypot.server'

export const headers: HeadersFunction = () => ({
	'Access-Control-Allow-Origin': '*',
	'Access-Control-Allow-Methods': 'POST',
	'Access-Control-Allow-Headers':
		'Content-Type, Authorization, X-Requested-With',
	'Access-Control-Allow-Credentials': 'true',
})

export async function action({ params, request }: ActionFunctionArgs) {
	const { listId } = params
	const formData = await request.formData()
	const domain = getEnv().DOMAIN
	const list = await prisma.list.findUnique({
		where: { id: listId },
		select: { id: true, signUpPage: { select: { redirectUrl: true } } },
	})

	invariantResponse(list, 'List not found', { status: 404 })
	checkHoneypot(formData)

	const submission = await parseWithZod(formData, {
		schema: EmailSchema.superRefine(async (data, ctx) => {
			const sub = await prisma.sub.findFirst({
				where: {
					AND: [{ email: data.email }, { listId: list.id }],
				},
			})

			if (sub) {
				ctx.addIssue({
					path: ['email'],
					code: z.ZodIssueCode.custom,
					message: 'Already subscribed',
				})
				return
			}
		}),
		async: true,
	})

	if (submission.status !== 'success') {
		console.log('error', submission.reply().error?.email)
		return json({
			res: {
				error: submission.reply().error?.email,
				redirect: undefined,
			},
		})
	}

	const { email } = submission.value

	const sub = await prisma.sub.create({
		data: {
			email,
			shortId: nanoid(8),
			list: { connect: { id: listId } },
		},
		select: {
			shortId: true,
		},
	})

	const redirect = list.signUpPage?.redirectUrl
		? list.signUpPage?.redirectUrl
		: `${domain}/me/${sub.shortId}`

	return json({ res: { error: undefined, redirect } })
}
```

When a user attempts to sign up via one of our embedded forms on external websites, we encounter a CORS error, which is expected. However, during sub creation—a process that functions seamlessly throughout the rest of the application—we face a PrismaClientRustPanicError, causing the entire application to crash until we manually restart it.

![Screenshot 2024-03-08 at 20 18 06](https://github.com/epicweb-dev/epic-stack/assets/84042097/91be6712-d5a8-4aaa-9828-e2a682b57584)

Does anyone have insights into what might be causing this problem?

**Comment by ludralph** [comment]
Hi @maxjnq 👋 

Thank you for raising this question.

When you catch a `PrismaClientRustPanicError`, it means that the underlying Prisma engine has crashed. In this case, you should restart the whole Node process, as the crash may have affected other parts of your application as well. For what is causing the crash, it's not really clear based on the information you provided, can you please enable [debugging](https://www.prisma.io/docs/orm/prisma-client/debugging-and-troubleshooting/debugging) so we can get access to more logs and share them here.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
